marshaling_rules.md

```diff
@@ -0,0 +1,121 @@
+| `Strict`   | `IteratorOnly` | `Either<&[Rint], AltRepInt>` |
+| `Relaxed`  | `UnfoldToVec`  | `Vec<Rint>`; `&[Rint]`; `Either<&[Rint], AltRepInt>`; `Vec<i32>`; `&[i32]`; `Either<&[i32], AltRepInt>` |
+| `Relaxed`  | `IteratorOnly` | `Either<&[Rint], AltRepInt>` ; `Either<&[i32], AltRepInt>`|
+| `Runtime`  | `UnfoldToVec`  | Any |
+| `Runtime`  | `IteratorOnly` | `Either<&[Rint], AltRepInt>` ; `Either<&[i32], AltRepInt>` |
+
+# Vector Types
+## Terminology
+A 'vector' is a primitive type used in `R`. Vectors are designed to behave as a strongly typed 1D array of objects. There are two different implementations of vector types: one is basically a pointer to a contiguous block of memory with known length (and some additional metadata), another is an iterator deigned to store rules for generating sequences of elements (instead of storing potentially very large vectors in memory). Array-based vectors shall be referred to as 'plain old data' (POD), iterators -- as ALTREP.
+
+`R` recognizes the following vector types that are directly exposed to the user:
+ - `logical (i32)`
+ - `integer (i32)`
+ - `real (f64)`
+ - `complex (f64, f64)`
+ - `raw (u8)`
+ - `character (usize)` (collection of pointers to character arrays)
+
+Each vector can contain special `NA` values. None of the primitive types have built-in support for `NA` (including `f64`, which has notion of `NaN`, a different thing), so `R` treats one value from the range of allowed values as `NA`. For instance, `NA_integer_` is `i32::MIN`, which is `1i32 << 31 = -2147483648`. As a result, `x <- -2147483648L` results in an error in `R`.
+
+## `Rust` counterparts
+
+`R` objects passed to `Rust` require additional validation and transformation. Let us define the following types:
+- `struct Rint(i32)`
+- `struct Rbool(i32)`
+- `struct Rfloat(f64)`
+- `struct Rbyte(u8)`
+- `struct Rcmpl(f64, f64)`
+- `struct Rstr(usize)` (?)
+
+Each of these types is binary compatible with their underlying type. An array of, say `i32`, represented by a `*i32` pointer and length, can be reinterpreted as `*Rint` of the safe length. 
+This can be the preferred solution when dealing with `R` plain vectors.
+
+### ALTREP
+**TODO: make a detailed description of ALTREP**
+Assume there are ALTREP wrappers for each type
+- `AltRepInt`
+- `AltRepBool`
+- `AltRepFloat`
+- `AltRepByte`
+- `AltRepCmpl`
+- `AltRepStr`
+
+Altreps should be iterators of types `Rxxx`, so `AltRepInt` should produce `Rint`.
+If altreps cannot produce `NA`s, then they can iterate over primitive types like `i32` (i.e. with no `NA` validation mechanism).
+
+**Bare ALTREP types should not be allowed as function parameters, but may be used as return types**.
+
+### Handling both plain and ALTREP vectors
+To enable direct consumption of ALTREP vector wrappers (without explicit memory allocations), a simple approach can be used. The monad `Either` is an excellent solution for representing either plain vector or its ALTREP wrapper.
```

Comment:
Why not? Not asking to expose `R` implementation of ALTREP, but a `Rust` wrapping iterator that only produces values?

I read a little bit about altreps, if I understood correctly, it is more or less an interface, altreps can represent not only simple things like sequences but also data stored on disk that cannot fit in memory (or should be read on demand).

If that is so, imagine a dot product method that accepts two vectors.

If it is a
```Rust
fn dot_product(x : Either<&[RFloat], AltRepFloat>, y  : Either<&[RFloat], AltRepFloat>) -> f64
```
then developer can decide at runtime if they are dealing with pre-allocated blocks of memory, which can be potentially read in chunks, verified against `NA` in chunks, and their product can be computed using some fuse-multiply-add and other intrinsic tricks, or if they are dealing with vector generators, which potentially wrap a very large data array with extremely poor support of random access. 

Of course, it is an advanced case, but I will not be surprised if such things can outperform some basic implementations of `R` functions, or be extremely memory-efficient.


At the same time, for general use-cases, just accept some `Vec<f64>`, and your altreps will be unfolded and each element verified against `NA` using simple checks. A lot of overhead but simple syntax, good for prototyping and `{rextendr}`.

Reply:
For example, cpp11 handles it internally and the users don't need to care whether the input is an ALTREP or not. I feel this is easy to use.

https://github.com/r-lib/cpp11/blob/5fb3137c62c773b224bf1b3f5092bd556ea20db3/inst/include/cpp11/doubles.hpp#L30

Reply:
I know, it is a reasonable solution, but afaik  cpp11 does not allow for `double[]` or `int[] `. How should we treat `Vec<T>` in extendr?

Reply:
> How should we treat `Vec<T>` in extendr?

I think I agree with you on this. We can make sure the vector is in the unfolded form and construct a `Vec<T>` or `&[T]` from the raw memory. That's totally fine.

What I'm suggesting is something like `Doubles` instead of `Either<&[RFloat], AltRepFloat>`.

Reply:
Honestly, I care little about performance and I'll probably stick with `Vec<T>` or `&[T]` for the sake of simplicity, so I'm not the right person to discuss this... Maybe @andy-thomason has some thoughts?

Reply:
> What I'm suggesting is something like `Doubles` instead of `Either<&[RFloat], AltRepFloat>`.

I think that's an implementation detail, as `Doubles` could just be a wrapper for `Either`. However,  `Doubles` is preferable in my opinion, because it would present a single interface, so downstream users don't have to constantly special-case their code.